capsules/src/alarm.rs

```diff
@@ -148,43 +145,40 @@ impl<'a, A: Alarm> Driver for AlarmDriver<'a, A> {
                             self.num_armed.set(self.num_armed.get() + 1);
                         }
 
-
-                        let now = self.alarm.now();
-                        td.t0 = now;
                         td.expiration = Expiration::Abs(time as u32);
 
                         if self.alarm.is_armed() {
                             (ReturnCode::SuccessWithValue { value: time }, true)
                         } else {
-                            self.alarm.set_alarm(time as u32);
-                            (ReturnCode::SuccessWithValue { value: time}, false)
+                            //self.alarm.set_alarm(time as u32);
+                            (ReturnCode::SuccessWithValue { value: time}, true)
                         }
                     },
                     _ => (ReturnCode::ENOSUPPORT, false)
+                };
+                if reset {
+                    self.reset_active_alarm(now);
                 }
+                return_code
+            })
+            .unwrap_or_else(|err| match err {
+                Error::OutOfMemory => ReturnCode::ENOMEM,
+                Error::AddressOutOfBounds => ReturnCode::EINVAL,
+                Error::NoSuchApp => ReturnCode::EINVAL,
             })
-            .unwrap_or_else(|err| {
-                let e = match err {
-                    Error::OutOfMemory => ReturnCode::ENOMEM,
-                    Error::AddressOutOfBounds => ReturnCode::EINVAL,
-                    Error::NoSuchApp => ReturnCode::EINVAL,
-                };
-                (e, false)
-            });
-        if reset {
-            self.reset_active_alarm();
-        }
-        return_code
     }
 }
 
+fn past_from_base(cur: u32, now: u32, prev: u32) -> bool {
```

Comment:
Can you rename this function something like `current_alarm_has_expired`? This name makes me confused!